docs/transform.ipynb

```diff
@@ -99,15 +137,271 @@
+    "    batch_replace = []\n",
+    "    for i, op in circuit.findall_operations(lambda op: op.gate == self._source):\n",
+    "      batch_replace.append((i, op, self._target.on(*op.qubits)))\n",
+    "    transformed_circuit = circuit.unfreeze(copy=True)\n",
+    "    transformed_circuit.batch_replace(batch_replace)\n",
+    "    return transformed_circuit\n",
+    "\n",
+    "q = cirq.NamedQubit(\"q\")\n",
+    "circuit = cirq.Circuit(cirq.X(q), cirq.CircuitOperation(cirq.FrozenCircuit(cirq.X(q), cirq.Y(q))), cirq.Z(q))\n",
+    "substitute_gate = SubstituteGate(cirq.X, cirq.S)\n",
+    "print(\"Original Circuit:\", circuit, \"\\n\", sep=\"\\n\")\n",
+    "print(\"Reversed Circuit:\", reverse_circuit(circuit), \"\\n\", sep=\"\\n\")\n",
+    "print(\"Substituted Circuit:\", substitute_gate(circuit), sep=\"\\n\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "OLJrXdyTCGev"
+   },
+   "source": [
+    "## `cirq.TransformerContext` to store common configurable options. \n",
+    "You might be wondering what's the benefit of adding the `@cirq.transformer` decorator on the methods / classes as defined above? Does it add any extra functionality to the decorated objects at all? \n",
+    "\n",
+    "Yes! The transformer decorator inspects the `cirq.TransformerContext` argument and automatically appends useful functionality like support for automated logging and recursively running the transformer on nested sub-circutis. \n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "5OuMLlNAFPcW"
+   },
+   "source": [
+    "### `cirq.TransformerLogger` and support for automated logging.\n",
+    "The `cirq.TransformerLogger` class is used to log the action of a transformer on an input circuit. Decorating a method / class with the `@cirq.transformer` decorator automatically adds support for logging. \n",
+    "\n",
+    "You can log the action of running a series of transformers on an input circuit by sharing the same logger instance across transformer calls. \n",
+    "\n",
+    "This is very useful to inspect the actions of individual transformers on the input circuit when running transformer bundles like `cirq.optimize_for_target_gateset`."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "1-1r6Ha9F1c3"
+   },
+   "outputs": [],
+   "source": [
+    "context = cirq.TransformerContext(logger=cirq.TransformerLogger())\n",
+    "transformed_circuit = reverse_circuit(circuit, context=context)\n",
+    "transformed_circuit = substitute_gate(transformed_circuit, context=context)\n",
+    "context.logger.show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "UvFdvxQ6MZD9"
+   },
+   "source": [
+    "### Built-in support for `deep=True`.\n",
+    "Notice that the custom transformers used above only modified the top-level circuit and had no impact on sub-circuits wrapped inside `cirq.CircuitOperation`.\n",
+    "\n",
```

Comment:
Maybe add a code cell to further clarify this difference via some output comparison ?

Reply:
Restructured, this is moot now.